sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -4782,6 +4819,13 @@ case class GpuOverrides() extends Rule[SparkPlan] with Logging {
         newPlan = GpuOverrides.fixupCpuReusedExchanges(newPlan)
       }
 
+      // If Hybrid Scan is enabled, we need to decide whether to push down a filter to the CPU
+      // or not according to whether CPU/GPU support it. After that we can remove the condition
+      // from one side to avoid duplicate execution or unnecessary fallback/crash.
+      if (conf.useHybridParquetReader) {
```

Comment:
It seems that this transformation only works when AQE is ON. Should it be inserted into `GpuTransitionOverrides` rather than here? And then, we can do pattern matching which targets on `GpuFilterExec` and `HybridFileSourceScanExec` directly. So, we can get rid of the potential unexpected cases that `FilterExec` fallbacks to CPU at last.  However,  to modify that data filters of `HybridFileSourceScanExec`, a new method has to be added.  

In addition, this sub rule shall be applied before `insertCoalesce` in case inserting the following `GpuCoalesceBatch` for the `GpuFilterExec` to be removed.

Reply:
> It seems that this transformation only works when AQE is ON.
Ok will move out of it.

> So we can finally get rid of the potential unexpected cases where FilterExec fallbacks to the CPU.

What kind of unexpected cases do you consider FilterExec fallbacks to be? In current code, if a condition is not supported by either CPU or GPU, it should fallback in FilterExec.

I think in `GpuTransitionOverrides` if an expression is not supported by GPU then the FilterExec already fallback, but we could push it to cpu to avoid fallback. So we also need to pattern match `HybridFileSourceScanExec => C2R => FilterExec`, do filter split and convert it to `new HybridFileSourceScanExec => new GpuFilter`, which seems unnatural to me.

Reply:
"if an expression is not supported by GPU then the FilterExec already fallback, but we could push it to cpu to avoid fallback"

That is very good point!